### DISCUSSION TITLE
Help with confuse (config) error in beets ydl plugin?

### DISCUSSION THREAD

**Discussion by reagle** [beetbox/beets#4615] [Q&A] [answered]
Hi, just thought I'd try beets and its ydl plugin but it looks like people have encountered vmassuchetto/beets-ydl#5 and the package is no longer maintained.

Looking at the code/error, it appears confuse is a configuration library, which I'm unfamiliar with, and I don't know why the following generates an error.

```
params.get("logtostderr", False)
```

At first I thought params was a dictionary, but it is apparently a `<class 'confuse.core.Subview'>` and I don't know what parameters that class accepts.  Any pointers?

**Comment by reagle** [comment]
The error in question is below. Reviewing the confuse documentation, it seems as if this is using an old API? Or from the current documentation, it doesn’t seem as if you pass a keyname as a parameter to get, rather you specify it by way of dictionary bracket syntax?

“””
File "/usr/lib/python3.8/site-packages/youtube_dl/YoutubeDL.py", line 353, in __init__
    self._screen_file = [sys.stdout, sys.stderr][params.get('logtostderr', False)]
TypeError: get() takes from 1 to 2 positional arguments but 3 were given
“””

**Comment by jackwilsdon** [comment, marked as answer]
It looks like the `YoutubeDL` class is expecting a dictionary as its argument, but it is instead being given a confuse subview. I'm not quite sure how this has ever worked, as subview's get function has never accepted more than 1 argument.

Confuse provides a flatten function which you could use to convert the subview to a normal dictionary (or in this case an ordered dictionary, but they should be roughly compatible). Try changing this line:

https://github.com/vmassuchetto/beets-ydl/blob/020bb138c2ce1b26f7a886f2f479f3ed149ac9f1/beetsplug/ydl.py#L153

To use the flattened version:

```
y = YoutubeDL(youtubedl_config.flatten())
```

**Reply by reagle** [reply]
Thank you, I was wondering if the API changed, but apparently not. Your suggestion, however, is on point, thank you!

**Reply by reagle** [reply]
BTW: The flatten might then cause later problems.

```
❯ beet ydl https://www.youtube.com/watch\?v\=VHoT4N43jK8
Traceback (most recent call last):
  File "/Users/reagle/.pyenv/versions/3.11.0/bin/beet", line 33, in <module>
    sys.exit(load_entry_point('beets==1.6.0', 'console_scripts', 'beet')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.11.0/lib/python3.11/site-packages/beets/ui/__init__.py", line 1285, in main
    _raw_main(args)
  File "/Users/reagle/.pyenv/versions/3.11.0/lib/python3.11/site-packages/beets/ui/__init__.py", line 1272, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/Users/reagle/.pyenv/versions/3.11.0/lib/python3.11/site-packages/beetsplug/ydl.py", line 105, in ydl_func
    outer_class.youtubedl(lib, opts, arg)
  File "/Users/reagle/.pyenv/versions/3.11.0/lib/python3.11/site-packages/beetsplug/ydl.py", line 191, in youtubedl
    y = YoutubeDL(youtubedl_config.flatten())
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.11.0/lib/python3.11/site-packages/youtube_dl/YoutubeDL.py", line 473, in __init__
    pp_class = get_postprocessor(pp_def_raw["key"])
                                 ~~~~~~~~~~^^^^^^^
TypeError: string indices must be integers, not 'str'
```

Given

```
directory: ~/f/mp3/1/
library: ~/f/mp3/beet-library.db
import:
    copy: no
plugins: ydl

# https://github.com/vmassuchetto/beets-ydl
ydl:
    download: True         # download files from sources after getting information,
    split_files: True      # try to split album files into separate tracks,
    import: True           # import files on youtube-dl after downloading 
                           #   and splitting,
    urls: []               # list of default urls to download when no arguments 
                           #    are provided, you can provide a playlist to 
                           #    get checked every time
    # youtubedl_options: {}     # youtube-dl available options -- https://git.io/fN0c7
    youtubedl_options:     # youtube-dl available options -- https://git.io/fN0c7
        format: 'best'
        postprocessors:
            key: 0
            preferredcodec: 'mp3'
            preferredquality: '320'
            nopostoverwrites: True
```
